google-cloud-spanner/lib/google/cloud/spanner/interval.rb

```diff
@@ -0,0 +1,176 @@
+            when HOURS
+              interval_nanoseconds += Interval.hours_to_nanoseconds numeric_value
+            when MINUTES
+              interval_nanoseconds += Interval.minutes_to_nanoseconds numeric_value
+            when SECONDS
+              interval_nanoseconds += Interval.seconds_to_nanoseconds numeric_value
+            end
+          end
+
+          Interval.new interval_months, interval_days, interval_nanoseconds
+        end
+
+        def self.years_to_months years
+          Integer(years) * 12
+        end
+
+        def self.hours_to_nanoseconds hours
+          Integer(hours) * NANOSECONDS_IN_AN_HOUR
+        end
+
+        def self.minutes_to_nanoseconds minutes
+          Integer(minutes) * NANOSECONDS_IN_A_MINUTE
+        end
+
+        def self.seconds_to_nanoseconds seconds
+          # We only support up to nanoseconds of precision
+          split_seconds = seconds.split "."
+          if split_seconds.length > 2 && split_seconds[1].length > 9
+            raise ArgumentError, "The seconds portion of the interval only supports up to nanoseconds."
+          end
+
+          Float(seconds) * NANOSECONDS_IN_A_SECOND
+        end
+
+        def self.from_months months
+          Interval.new months, 0, 0
+        end
+
+        def self.from_days days
+          Interval.new 0, days, 0
+        end
+
+        def self.from_seconds seconds
+          nanoseconds = Interval.seconds_to_nanoseconds seconds
+          Interval.new 0, 0, nanoseconds
+        end
+
+        def self.from_nanoseconds nanoseconds
+          Interval.new 0, 0, nanoseconds
+        end
+
+        def to_s
+          years = 0
+          months = 0
+          days = @days
+          hours = 0
+          minutes = 0
+          seconds = 0
+          remaining_nanoseconds = @nanoseconds
+
+          years = @months / 12
+          months = @months % 12
+          hours = Integer(remaining_nanoseconds / 3_600_000_000_000)
+          remaining_nanoseconds %= 3_600_000_000_000
+          minutes = Integer(remaining_nanoseconds / 60_000_000_000)
+          remaining_nanoseconds %= 60_000_000_000
+          seconds = remaining_nanoseconds / 1_000_000_000
+
+          interval_string = "P"
+
+          if years != 0
+            interval_string += "#{years}Y"
+          end
+
+          if months != 0
+            interval_string += "#{months}M"
+          end
+
+          if days != 0
+            interval_string += "#{days}D"
+          end
+
+          if hours != 0 || minutes != 0 || seconds != 0
+            interval_string += "T"
+
+            if hours != 0
+              interval_string += "#{hours}H"
+            end
+
+            if minutes != 0
+              interval_string += "#{minutes}M"
+            end
+
+            if seconds != 0
+              if (seconds % 1).zero?
+                interval_string += "#{Integer(seconds)}S"
+              else
+                interval_string += "#{seconds}S"
+              end
```

Comment:
Instead of doing all that (if you're worried about `seconds` being a float) just use `"#{seconds.to_i}S"`

However, don't we need to output the nanoseconds component somehow?

Reply:
There is no nanoseconds component in the ISO8601, but the seconds component can be a decimal point value, hence why I convert it to a fractional.

I did it this way because we are using a float for seconds it can print "1.0" and we just need the "1", but if it does have a decimal point, we want to print it as well. But I guess I could change the first part of the if to `seconds.to_i`.

Reply:
seconds will always be an integer at this point. (In line 216 you're using an integer divide, and the remaining nanoseconds are being thrown away.) You can either (1) convert to a float on line 216 by saying `seconds = remaining_nanoseconds.to_f / 1_000_000_000` or (2) take the remainder of that calculation as remaining_nanoseconds and render that here explicitly if nonzero.